# Citations

In this section of the book you will learn how to include bibliographies into your document, how to generate citations or source them from external places. We will touch on [RefMangeR](https://github.com/ropensci/RefManageR/) and [bibtex](https://cran.r-project.org/web/packages/bibtex/index.html) packages to help us with references. In addition we will also learn [citr](https://github.com/crsh/citr) package. Finally we will mention external software that can help you manage large number of citations.

For this section of the book we need to install these additional packages

```{r, eval = F}
install.packages(c("bibtex", "RefManageR"))
```

## Introductions

[A citation is a reference to a source (from wikipedia)](https://en.wikipedia.org/wiki/Citation). R provides nice function `citation()` that helps us generating citation blob for R packages that we have used

Let's try generating citation text for [rmarkdown](https://github.com/rstudio/rmarkdown) package by using the following command

```{r, comment=""}
citation("rmarkdown")
```

--- great

## BibTex {#bibtex}

### Introduction

[BibTex](http://www.bibtex.org/Format/) is one of many citation formats, [this is a more comprehensive list](https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#Specifying_a_Bibliography) of all possible citation formats. Rmarkdown works best with [BibTex](http://www.bibtex.org/Format/) file format and this is the one we are going to cover today. Typical file extention for BibTex is `.bib`

Let's try converting our citation into bibtex format using the following code

```{r, comment=""}
#toBibTex(citation("rmarkdown"))

citation("rmarkdown") %>% toBibtex()
```

### `knitr::write_bib` {#make-refs}

We have seen in the previous section how we can generate citation and `.bib` format. In order for us to start using those citations within our Rmarkdown documents we need them to be saved into a `.bib` file. For that we are going to use `write_bib()` function from [knitr package](https://cran.r-project.org/web/packages/knitr). 


```{r}
knitr::write_bib("rmarkdown", file = "references.bib")
```

### BibTex keys {#bib-keys}

It is a bit of a round trip, for some read `citation()` function in [BibTex section above](#bibtex) doesn't make keys for our bibtex entries, whereas `knitr::write_bib()` function does. Let's read our bibtex file back in using the following code to see what our keys are. We will learn how to use those key in the [Inline citation section below](#inline-cite)

```{r, comment=""}
library(bibtex)

refs <- read.bib("references.bib")
keys <- names(refs)
keys
```

Let's open our `references.bib` file by clicking on it in the "File" pane-tab and compare that bibtex with the one in [BibTex section above](#bibtex). Can't you find location of those keys in the bibtex file?

```{r echo = F, comment=""}
toBibtex(refs)
```

## Where do I get `.bib` file?

### Citation managment software

This a short list of external software that may help you with managing citation. Those four will have an option to export a list of citation as a `.bib` file. I will leave for you as a home to figure out how to export `.bib` file from your citation manager of choice.

- [Zotero](https://www.zotero.org/)
- [Paperpile](https://paperpile.com)
- [Mendeley](https://www.mendeley.com)
- [EndNote](https://endnote.com/)

### ZoteroBib service

> [zbib.org](https://zbib.org/)

Zotero provides nice free service for getting citation and exporting them as `.bib` files 

### RefMangeR

The [RefManageR](https://cran.r-project.org/web/packages/RefManageR/index.html) maybe a useful package to try to source external references

```{r eval = F}
library(RefManageR)

#keywords <- c("FastQC",
#              "BWA-MEM",
#              "featureCounts",
#              "RNAsik")

#citations <- lapply(keywords, ReadCrossRef, limit = 1)

titles <- c("FastQC: A Quality Control Tool for High Throughput Sequence Data",
            "Aligning Sequence Reads, Clone Sequences and Assembly Contigs with BWA-MEM",
            "featureCounts: An Efficient General Purpose Program for Assigning Sequence Reads to Genomic Features",
            "RNAsik: A Pipeline for Complete and Reproducible RNA-Seq Analysis That Runs Anywhere with Speed and Ease")

citations <- lapply(titles, ReadCrossRef, limit = 1)

citations

lapply(citations, WriteBib, file = "references2.bib", append = TRUE)
```

### Journal websites

A lot of journal website as well as [Google Scholar](https://scholar.google.com/) will have some export citation button. Look for those "export" button and select `.bib` file format

## Linking `.bib` file with Rmarkdown document

Let's practice linking bibtex file with our Rmarkdown document via YAML header. We are going to use `references.bib` file that we have generated in the [section above](#make-refs). Start by including the following key value pair into your YAML header

```YAML
---
bibliography: "references.bib"
---
```

You can include multiple reference files using the following syntax, alternatively you can concatenate two bib files into one.

```YAML
---
bibliography: ["references1.bib", "references2.bib"]
---
```

## Inline citation {#inline-cite}

Now we can start using those bib keys that we have learned [just before](#bib-keys), using the following syntax

- `[@key]` for single citation
- `[@key1; @key2]` multiple citation can be separated by semi-colon
- `[-@key]` in order to suppress author name, and just display the year
- `[see @key1 p 12; also this ref @key2]` is alos a valid syntax


Let's start by citing the [rmarkdown package](https://cran.r-project.org/web/packages/rmarkdown/index.html) using the following code and press `Knit` button. I am also including a new first level header and a short sentence.

```{r eval = F}
# Citation section

I have been using Rmarkdown package [@R-rmarkdown]. It is pretty cool. I should also go and read [@rmarkdown2018; and @rmarkdown2020] books. 
```

## References section

As you can see references list will be added to the very end of the document. If add a new header called "References" to the bottom of our document that will section our references list out. Let's see what happens when we include first level header at the bottom of the document and press `Knit`, use the following code

```markdown
# References
```

Alternatively we can use the following code to include references anywhere in the document. Let's try puting our references list at the very top, just to that we can.

```markdown
# References

<div id="refs"></div>
```

--- funny 

## Citation style

> [list of csl files on zotero](shttps://www.zotero.org/styles)

Most likely you will need to use specific style for your citation. [Citation Style Language (CSL)](https://citationstyles.org/) is the language used to define the style in the `.csl` file. You need to download the `.csl` file for your desired style and link it with Rmarkdown document via YAML header. [Zotero's website](https://www.zotero.org/styles) is the most comprehensive collection of styles, but there are other places to source `.csl` files, for example [this github repository](https://github.com/citation-style-language/styles).

Let's programmatically download [Nature's journal](https://nature.com) citation style from [zotero](shttps://www.zotero.org/styles), link it with our Rmarkdown document and `Knit` using the following code

```{r, eval = F}
download.file("https://www.zotero.org/styles/nature?source=1", destfile = "nature.csl")
```

```YAML
---
csl: nature.csl
---
```

## Linking inline citation with the references list

You can also enable internal linking of the citation to the corresponding entry in the references list using the following YAML key

```YAML
---
link-citations: true
---
```

## Footnotes

```{r, eval = F}
^[Footnote goes here]
```

##  Citr addin

https://github.com/crsh/citr
